---
title: "CORALIS-vignette"
output: 
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
    math: katex
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{CORALIS-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
<style>
body {
text-align: justify}
</style>

```{r setup, include = FALSE, message=FALSE}
require(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
opts_knit$set(root.dir = normalizePath('../'))

#Loading PKG
library(knitr)
library(kableExtra)
library(stringr)
library(dplyr)
require(prettydoc)
```

# 1. CORALIS

CORALIS is an R package for non-coding RNA target enrichment analysis exclusively based on experimentally validated interactions. This package allows for the identification of enriched target genes given a list of ncRNAs defined by the user. To do so, CORALIS access to [miRTarbase](http://mirtarbase.cuhk.edu.cn/php/index.php) and [RNAInter database](http://www.rna-society.org/rnainter/) and retrieves records associated with experimentally supported interactions between ncRNAs and their gene targets. So far, CORALIS supports miRNA, lncRNA, snRNA and snoRNA target enrichment analysis for several species such as *Homo sapiens*, *Rattus norvegicus*, *Caenorhabditis elegans* and *Drosophila melanogaster*, among others. 


# 2. State of Art

Non-coding RNAs (ncRNAs) are RNA molecules involved gene expression regulation by interacting and modulating the complex molecular and cellular processes. The combination of RNA large-scale sequencing and computational analyses has revealed the existence of a vast diversity of RNA species with essential roles in biological functions. In accordance with their sequence length, ncRNAs are classified into two groups: small ncRNAs (sncRNAs) (<200 nts) such as snoRNAs, snRNAs, miRNAs, etc.; and long ncRNAs (ncRNAs) (>200 nts) [1]. Whereas in the past ncRNAs were considered as intracellular waste, nowadays they are under the spotlight for researchers since they have been found to be involved in the modulation of relevant signalling pathways and almost all molecular processes by altering chromatin modification, gene expression and protein-protein interactions, among others.  

Therefore, the accumulating evidence show us that the ncRNA analysis has become a challenge, with an urgent need of a more sensitive and comprehensive ncRNA annotation and functional interaction analysis. The available information of ncRNA-gene target interactions is stored in multiple different databases and data warehouses. In parallel with the development of new target interaction prediction algorithms, the amount of interactions listed in those repositories have grown enormously. Dealing with this huge network of interactions becomes a much harder task due to the nature of ncRNA-target interactions in which a single ncRNA blocks the expression of multiple gene, and in turn, a single gene can be targeted by several ncRNAs.


Here we present **CORALIS**, an R package focused on the statistical analysis of ncRNA-target validated interactions and visualization. CORALIS allows the identification of potential target genes for a variety of ncRNAs for different species.


**Table 1** shows the type of ncRNAs and species supported in **CORALIS**, as well as the number of experimentally validated interactions. 


```{r Table_1, echo=FALSE}
# List of available interactions with experimental support by species
table_nums <- captioner::captioner(prefix = "Table")
tab.1_cap  <- table_nums(name = "Table 1", 
                         caption = "Supported species in CORALIS for ncRNA-gene target enrichment analysis. Each entry indicates the number and type of ncRNA-target interaction in CORALIS by specie.")
availsp() %>% 
  kbl(caption = tab.1_cap) %>%
  kable_classic(full_width = F, html_font = "Cambria")
```


# 3. CORALIS ncRNA-target enrichment statistics 

CORALIS contains a set of functions that calculates the enriched target genes from a list of ncRNAs provided by the user (input). CORALIS builds a 2x2 contingency table for each target gene and executes an hypergeometric distribution test, commonly known as one-tailed Fisher’s exact test. Table 2 shows the 2x2 contingency table.


```{r Table2, echo = FALSE}
# Build 2x2 matrix
mat<-matrix(0L, nrow = 3, ncol = 3)
colnames(mat)<-c(" ncRNAs in input ", "NOT ncRNAs in input","Total")
rownames(mat)<-c(" gene i", " NOT gene i", "Total")
mat[,1]<-c("a","c","a+c")
mat[,2]<-c("b","d","b+d")
mat[,3]<-c("a+b","c+d","N=a+b+c+d")
df<-as.data.frame(mat)

# Setup title
tab.2_cap  <- table_nums(name = "Table 2", 
                         caption = "The 2X2 contingenc table for gene i analysis as potential gene-target of ncRNAs that user set in the input with the cell fequencies represented as a, b, c and d, and the marginal totals as a+b, c+d, a+b, and c+b")

# print table in pretty format
df %>% 
    kbl(caption = tab.2_cap) %>%
    kable_classic(full_width = F, html_font = "Cambria") 

```

\n 
where *i* is a gene target for any ncRNA in the input with annotation in the source databases (miRTarbase for miRNAs, or RNAInter database for lncRNAs, snoRNAs and snRNAs); *a* represents the hits, that is the number of ncRNAs introduced by the user in the input which interacts with gene *i*; *b* points to the number of ncRNAs in the source databases that target gene i but does not appear in the user's input; *a+b* is the total number of ncRNAs in the source databases that interacts with gene *i*; *c* represents the number of ncRNAs in the input that do not interact with gene *i* according with the information stored in the source databases; *d* is the number of ncRNAs in the source databases but not in the user's input that do not target gene *i*; *c+d* is the number of total known ncRNAs that do not interact with gene i; *N* points to the total number of ncRNAs annotated in the source databases.


The p-value is computed as follows:

$$p=1-\sum_{n = 0}^{a-1} \frac{\binom{a+b}{a} \binom{c+d}{c}}{\binom{N}{a+c}}$$

Moreover, CORALIS provides other additional interesting statistical analyses such as  the odds ratio (OR), the relative risk (RR) –plus their respective confidence interval ls at alpha=0.05 and standard errors - and the adjusted p-value (false discovery rate (FDR), using Benjamin-Hochberg correction [2]):
 
$$OR=\frac{(\frac{a}{b})}{(\frac{c}{d})}$$

# 4. Installation and package loading 

You can install both the released and the development version of CORALIS from [GitHub](https://github.com/Daniel-VM/CORALIS) with:

```{r Installation, eval=FALSE}
# install.packages("devtools")
devtools::install_github("Daniel-VM/CORALIS")

```


To be able to use miRNnode's functions we first have to load CORALIS:

```{r Load_library, eval=FALSE}
library(CORALIS)
```



# 5. ncRNA-target enrichment analyisis with CORALIS

The *tienrich()* function in CORALIS allows target enrichment analysis given a list of ncRNAs. This function scans the miRTarbase and the RNAInter databases, reporting those interactions with experimental support. It subsequently applies an hypergeometric distribution test to assess whether the target genes for the miRNA-input list are enriched.

* The following arguments in *tienrich()* must be defined by the user: 
  + *input_list*:  a vector of ncRNAs IDs. The IDs must be defined in the [miRBase](http://www.mirbase.org/search.shtml) format for microRNA-target enrichment analysis (ie: 'hsa-miR-3196') or [Official Gene Symbol](https://www.genenames.org/) format for the rest of ncRNAs (i.e: 'RUNX2'). 
  + *type*: character string indicating the type of ncRNA target enrichment analysis (this depends on the type of ncRNAs defined in the input). For example: miRNA_mRNA' for microRNA-target interactions, 'lncRNA_mRNA' for lncRNA-target interactions, 'snoRNA_mRNA' for snoRNA-target interactions or 'snRNA_mRNA' for snRNA-target interactions.
  + *min*: integer specifying the minimum number of interactions that target genes must have with the ncRNAs in the input in order to be analyzed. Default is min = 1.
  + *fdr*: the desired adjusted p-value threshold at which the enriched target-genes are found.
  + *organism*: character string specifying the organism name (i.e: "*Homo sapiens*"). Run *availsp()* to list all available species.
  

## 5.1 CORALIS example - Human microRNA dataset

Let’s imagine that we are analyzing the expression profile of ncRNAs, for example microRNAs (miRNAs), in a human specific cell type before and after certain treatment. Suppose for example that we obtain 9 out of 50 miRNAs with significantly higher expression levels after treatment. The next step would normally be to analyze the potential effect of these 9 deregulated miRNAs might promote on target genes and whether they belong to the same pathway. Therefore, **CORALIS** can be used to determine the potential target genes that may be enriched from an input list and their associated pathways.

Our aim here is to identify which genes can be potential enriched targets for these miRNAs. To do this we load an object containing 50 miRNAs IDs in ([miRBase id format](http://www.mirbase.org/search.shtml)). Among them, the 5 first items correspond to the deregulated miRNAs.

```{r load_dataset}
# Load rnasID object 
data("rnasID")

# rnasID contains the IDs of several ncRNAs (See: help(ids)). In this example we are going to choose 50 miRNAs where the first five have significantly higher expression levels after treatment (deferentially expressed miRNAs).
mirs<-ids[["miRNAs"]][1:9]
head(mirs)

```

This approach will allow us to use different miRNAs for the target enrichment from the same input list by selecting the number of miRNAs on each case. With this approach the user may explore different results for different thresholds of the input list (i.e., different FDR levels, different fold changes, etc.)

### 5.1.1 miRNA-target enrichment analysis

Before executing the target enrichment analysis it is recommended to check whether the input is in valid format. *tienrich()* uses as input a vector of IDs in which every item must be a string of characters. Let's apply some checks:

```{r input_check}
# check mirs class
class(mirs)

# check mirs format
is.vector(mirs)

```

Once the IDs have been checked, we are ready to carry out miRNA-target enrichment analysis on our set of human's miRNAs. The arguments used for the enriched target genes identification in this examples will be an adjusted p-value (FDR) <0.05 and at least 2 interactions with the miRNAs in the input. 

```{r tienrich}
# miRNA-target enrichment analysis with tienrich()
genetar <- tienrich( 
                    input_list = mirs, 
                    type = "miRNA_mRNA", 
                    organism = "Homo sapiens", 
                    min = 2, 
                    fdr = 0.05  
                    ) 


# Exploring tienrich output.
class(genetar)
genetar@results %>% head()
genetar@not_found

```

*Hint: By defining min = 2, only target genes with more than 2 interactions with miRNAs in the input will be candidates for target enrichment analysis (this argument is not mandatory).*


"*genetar*" stores the result of the miRNA-target enrichment analysis. First of all, *tienrich()* output is an object belonging to the **CoralisResult** class. This object contains 2 slots. 
* 1) The first one, @results, is the result of the miRNA-target enrichment analysis in which we have several information: i) the target gene in Official Gene Symbol format(Gene_symbol); ii) the number of interactions between the target gene and the ncRNAs in the input (num_int); iii) the ID of the ncRNAs involved in the interaction (ncRNAs); iv) the p-value (pvalue); v) the adjusted p-value (FDR); vi) the odds ratio (OR); vii) the standard-error of the OR (OR.SE); viii) the relative confidence intervals of the OR (OR.IC).

* 2)	The second slot, @notfound, stores the ncRNA IDs in the input list for which no experimentally validated target genes have been found.

### 5.1.2 Visualization of enriched targets

CORALIS also allows to visually analyze the output of the target enrichment analysis. Specifically, the *nodeVisu()* function provides a series of graphs to explore the interaction between the ncRNAs of interest and their target genes, as well as different statistical parameters obtained in the enrichment analyses.

Firstly, you can use the ncRNA-target enrichment output directly as input of *nodeVisu()*, but it also supports the slot @results independently. The user can choose the number of *top* target genes to be represented in the graphs.

With the *type* argument the graphic format is specified. 

* A) If *type* = "barplot" it creates a barplot for visualizing the number of interactions that top target genes have with the ncRNAs in the input. 

We will now visually explore the result of the miRNA-target enrichment analysis for the top 30 target genes. The following bar chart represents on the y-axis the name of the target genes, and on the x-axis the number of interactions that each of these genes has with the ncRNAs in the input. The gradient indicates the FDR of the enrichment analysis.

```{r Target_barplot}
# Barplot of targets genes
bar <- nodeVisu(obj = genetar,top = 30, type = "barplot")
print(bar)
```

* B) If *type* = "network", then it creates at html widget to analyze interactively the network of interactions between the ncRNAs and their top target genes according to the enrichment analysis.

This graphical representation provides a more detailed information on the interaction between ncRNAs and the top target genes selected. The nodes color of the 3D network graph represent all significant ncRNAs (miRNAs in this example) and their target genes in orange and purple, respectively. The radius size of the node indicates the number of interactions for one specific gene, that is, the larger the number of interactions between the enriched target gene and the ncRNAs in the input the larger will be the size of the node. On the other hand, the links/edges thickness reports the FDR of the interactions (the thicker the link, the lower the FDR, and therefore, the signification of the interaction is higher).

```{r Target_network}
# Interactive 3D network
net <- nodeVisu(obj = genetar, 
                top = 30, 
                fixedsize = FALSE, # if true, default node size is fixed
                type = "network")
net
```

### 5.1.3 Functional enrichment analysis

Once we have found the enriched target genes whose expression may be altered by the differentially expressed miRNAs, the next step is to identify the biological processes in which the enriched genes are involved, and could therefore be altered after the treatment.


There are available packages in Bioconductor that allow for functional enrichment analysis. Here we use the *enrichGO* function (*clusterProfiler* package [3]) to identify whether the top 25 gene targets are overrepresented in any biological process listed in the Gene Ontology database (see [Gene Ontholgy resources](http://geneontology.org/))[4].



```{r Functional_GOanalysis, message=FALSE}
# Load dependencies for pathway enrichment analysis

#if (!require("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#if (!require("biomaRt", quietly = TRUE))
#  BiocManager::install("biomaRt")
#library(biomaRt)
#if (!require("clusterProfiler", quietly = TRUE))
#  BiocManager::install("clusterProfiler")
#library(clusterProfiler)
#if (!require("org.Hs.eg.db", quietly = TRUE))
#  BiocManager::install("org.Hs.eg.db")
#library(org.Hs.eg.db)

# 1- convert gene target's symbol into ENTREZID format
gene <- genetar@results$Gene_symbol[1:25] # top 25 targets

# Use Homo sapiens dataset
ensembl <-useDataset("hsapiens_gene_ensembl",
                     mart = useEnsembl("ensembl","oaries_gene_ensembl"))

# Get the gene IDs
gene_id <- unique(getBM(attributes = c("hgnc_symbol", "entrezgene_id"),    
                        filters = "hgnc_symbol",
                        values = gene,
                        mart = ensembl))

# 2- Functional enrichment analysis (GO terms)
biopro <- enrichGO(gene          = gene_id$entrezgene_id,
                   keyType       = "ENTREZID",
                   OrgDb         = org.Hs.eg.db,
                   ont           = "BP",
                   pAdjustMethod = "fdr",
                   pvalueCutoff  = 0.05,
                   qvalueCutoff  = 0.05)

# Let's filter GO temrs which FDR (p.adjust) < 0.05
s_biopro <- subset(biopro@result, p.adjust < 0.05)
head(s_biopro, n = 10)

```

The functional analysis revealed 9 biological processes significantly altered by the deferentially expressed miRNAs after the treatment. Specifically the miRNAs enriched target genes ABCC1 (entrezID:4363), GRM5 (entrezID:2915) and ICAM1 (entrezID:3383) play role in [GO:1904646](https://www.ebi.ac.uk/QuickGO/term/GO:1904646). This GO term encompasses any process related to processes that results in alterations of cell activity as a consequence of a amyloid-beta stimulus.

```{r GOterm_inspect}
# Filter the miRNA target enrichment matrix to identify the miRNAs involved in the deregulation of genes in GO:1904646
genes_go1904646_entrezid <- subset(s_biopro, ID == "GO:1904646") %>%
                            dplyr::select(., geneID) %>%
                            str_split(., "/", simplify = FALSE) %>%
                            unlist(.)

genes_go1904646_symbol <- subset(gene_id, entrezgene_id %in% genes_go1904646_entrezid) %>% 
                          dplyr::select(., hgnc_symbol) %>% 
                          pull(.)

# Fin the miRNAs involved in the regulation of genes in GO:1904646
mirnas_go1904646<- subset(genetar@results, Gene_symbol %in% genes_go1904646_symbol)
mirnas_go1904646
```
We can now individually analyze whether the significant enriched target genes from the input list are significantly involved in a specific GO term, such as the GO:1904646.
This information can be depicted in  network of interactions where the genes within a specific GO term and their related miRNAs are represented.

```{r}
# Plot 3D network of interactions between genes in GO:1904646 and the deregulated miRNAs.
net_go1904646<- nodeVisu(obj = mirnas_go1904646,
                         type= "network")
net_go1904646
```
# 6. CONCLUSION

In conclusion, CORALIS helps to analyze the complex regulatory network between ncRNAs and their target genes. In this example we have carried out miRNA-target enrichment analysis in humans, but CORALIS supports other ncRNA-target enrichment analysis such lncRNAs, snRNAs or snoRNAs, as well as different species. Furthermore, CORALIS’s functionalities are enhanced when it is used in conjunction with other packages that allows functional analysis derived from the regulation of ncRNAs on their target genes. In further versions, CORALIS will include a wider range of ncRNAs, species, graphs and additional databases for interaction networks.

# 7. SESSION INFORMATION

```{r info_session}
sessionInfo()
```

# 8. REFFERENCES

1. Romano G, Veneziano D, Acunzo M, Croce CM. Small non-coding RNA and cancer. Carcinogenesis. 2017 May 1;38(5):485-491. doi: 10.1093/carcin/bgx026. PMID: 28449079; PMCID: PMC6248440.  

2. Benjamini, Y., and Hochberg, Y. (1995). Controlling the false discovery rate: a practical and powerful approach to multiple testing. Journal of the Royal Statistical Society Series B, 57, 289–300. http://www.jstor.org/stable/2346101.

3. Yu G, Wang L, Han Y, He Q (2012). “clusterProfiler: an R package for comparing biological themes among gene clusters.” OMICS: A Journal of Integrative Biology, 16(5), 284-287. doi: 10.1089/omi.2011.0118.

4. The Gene Ontology resource: enriching a GOld mine. Nucleic Acids Res. Jan 2021;49(D1):D325-D334.
